File: 2025/03/day3.py
#!/usr/bin/env python3
second = lambda x: x[1]

def joltage(bank, limit):
    offset = 0
    chosen = []
    for i in range(1-limit, 1):
        m = max(enumerate(bank[offset : i or None]), key=second)
        offset += m[0] + 1
        chosen.append(m[1])
    return int(''.join(chosen))

count1 = 0
count2 = 0

with open("input.txt") as f:
    for line in f:
        count1 += joltage(line.strip(), 2)
        count2 += joltage(line.strip(), 12)

print(count1)
print(count2)

File: 2025/03/day3.mjs
#!/usr/bin/env -S bun run
const inputs = await Bun.file("input.txt").text()

function maxdex(arr) {
  return arr.reduce((acc, val, idx) => {
    let [elem, maxidx] = acc
    if (val > elem) {
      elem = val
      maxidx = idx
    }
    return [elem, maxidx]
  }, [arr[0], 0])
}

function joltage(bank, limit) {
  const result = []
  let offset = 0
  let n = bank.length
  for (let stop = limit - 1; stop >= 0; --stop) {
    const slc = bank.slice(offset, n - stop)
    const [elem, idx] = maxdex(Array.from(slc))
    offset = offset + idx + 1
    result.push(elem)
  }
  return Number(result.join``)
}

console.log(
  inputs
    .split`\n`
    .map(n => [joltage(n, 2), joltage(n, 12)])
    .reduce((a, b) => [a[0] + b[0], a[1] + b[1]])
    .join`\n`
)
